### PR TITLE
SALTO-4993: Deploy assets object type continue

### PR DESCRIPTION
**Pull request by idoamit8**
In this PR I support order for object Types

---

_Additional context for reviewer_
None

---
_Release Notes_: 
_Jira Adapter:_
Now support deploy of order of assetsObjectTypes.

---
_User Notifications_: 
None

### CODE REVIEW

**Review by shir-reifenberg — COMMENTED** [review]
Great work! added some questions and comments

File: packages/jira-adapter/src/filters/assets/fetch_assets_object_type_order.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    objectTypes: { refType: new ListType(BuiltinTypes.NUMBER) },
+    assetsSchema: { refType: BuiltinTypes.NUMBER },
+  },
+  path: [JIRA, adapterElements.TYPES_PATH, ASSETS_OBJECT_TYPE_ORDER_TYPE],
+})
+
+const createAssetsObjectTypeOrder = (assetsObjectTypes: InstanceElement[], orderType: ObjectType): InstanceElement => {
+  const treeParent = assetsObjectTypes[0].value.parentObjectTypeId.value
+  const schema = assetsObjectTypes[0].annotations[CORE_ANNOTATIONS.PARENT]?.[0]
+  const name = `assetsObjectTypeOrder_${treeParent.elemID.name}`
```
[the remark this answers is not in the document]
Comment (idoamit8) [reply packages/jira-adapter/src/filters/assets/fetch_assets_object_type_order.ts:36]: I changed it a bit

File: packages/jira-adapter/src/filters/assets/deploy_assets_object_type_order.ts [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+const { awu } = collections.asynciterable
+const log = logger(module)
+
+const sendRequest = async ({
+  client, workspaceId, position, toObjectTypeId, assetsObjectType,
+}: {
+  client: JiraClient
+  workspaceId: string
+  position: number
+  toObjectTypeId?: string
```
[the remark this answers is not in the document]
Comment (idoamit8) [reply packages/jira-adapter/src/filters/assets/deploy_assets_object_type_order.ts:36]: The API wants what the API wants

File: packages/jira-adapter/src/filters/assets/deploy_assets_object_type_order.ts [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        deployResult: { appliedChanges: [], errors: [error] },
+        leftoverChanges,
+      }
+    }
+    const deployResult = await deployChanges(
+      relevantChanges.filter(isInstanceChange),
+      async change => {
+        const instance = getChangeData(change)
+        const parent = getParent(instance)
+        const toObjectTypeId = parent.elemID.typeName === ASSETS_OBJECT_TYPE ? parent.value.id : undefined
```
[the remark this answers is not in the document]
Comment (idoamit8) [reply packages/jira-adapter/src/filters/assets/deploy_assets_object_type_order.ts:81]: no. its a valid scenario that happen in the root object types

File: packages/jira-adapter/src/filters/assets/deploy_assets_object_type_order.ts [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+              await sendRequest({
+                client, workspaceId, position, toObjectTypeId, assetsObjectType,
+              })
+            })
+        }
+        if (isModificationChange(change)) {
+          const positionsBefore = change.data.before.value.objectTypes
+          await awu(instance.value.objectTypes).filter(isReferenceExpression)
+            .forEach(async (assetsObjectType, position) => {
+              if (positionsBefore[position]?.elemID.getFullName() !== assetsObjectType.elemID.getFullName()) {
```
[the remark this answers is not in the document]
Comment (idoamit8) [reply packages/jira-adapter/src/filters/assets/deploy_assets_object_type_order.ts:95]: talked F2F

File: packages/jira-adapter/src/filters/assets/assets_object_type_order.ts [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+              await deployOrderChange({
+                client, workspaceId, position, toObjectTypeId, assetsObjectType,
+              })
+            })
+        }
+        if (isModificationChange(change)) {
+          const positionsBefore = change.data.before.value.objectTypes
+          await awu(instance.value.objectTypes).filter(isReferenceExpression)
+            .forEach(async (assetsObjectType, position) => {
+              if (positionsBefore[position]?.elemID.getFullName() !== assetsObjectType.elemID.getFullName()) {
```
Comment (shir-reifenberg) [inline packages/jira-adapter/src/filters/assets/assets_object_type_order.ts:158]: maybe we should verify `positionsBefore` is what we think it is?

File: packages/jira-adapter/test/filters/assets/assets_object_type_order.test.ts [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+    position: id - 1,
+    parentObjectTypeId: new ReferenceExpression(parentObjectTypeInstance.elemID, parentObjectTypeInstance),
+  },
+  [JIRA, elementUtils.RECORDS_PATH, ASSESTS_SCHEMA_TYPE, 'assetsSchema', 'assetsObjectTypes', 'parentObjectTypeInstance', `assetsObjectType${suffix}`, `assetsObjectType${suffix}`],
+  {
+    [CORE_ANNOTATIONS.PARENT]: [new ReferenceExpression(assetSchema.elemID, assetSchema)],
+  }
+)
+
+describe('fetchAttributesFilter', () => {
```
Comment (shir-reifenberg) [inline packages/jira-adapter/test/filters/assets/assets_object_type_order.test.ts:47]: please change :)

File: packages/jira-adapter/test/filters/assets/assets_object_type_order.test.ts [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+  const assetsObjectTypeInstanceTwo = createAssetsObjectTypeInstance(2, 'Two', assetSchema, parentObjectTypeInstance)
+  const assetsObjectTypeInstanceThree = createAssetsObjectTypeInstance(3, 'Three', assetSchema, parentObjectTypeInstance)
+  describe('fetch', () => {
+    beforeEach(() => {
+      const config = _.cloneDeep(getDefaultConfig({ isDataCenter: false }))
+      config.fetch.enableJSM = true
+      config.fetch.enableJsmExperimental = true
+      filter = assetsObjectTypeOrderFilter(getFilterParams({ config, client })) as typeof filter
+    })
+    it('should add assetsObjectTypeOrder to the elements', async () => {
```
Comment (shir-reifenberg) [inline packages/jira-adapter/test/filters/assets/assets_object_type_order.test.ts:82]: Can we add a test to verify assetsObjectTypeOrder `ObjectType` is created as well ?

**Review by shir-reifenberg — COMMENTED** [review]
👑 
Added some comments mainly around safties :)

File: packages/jira-adapter/src/filters/assets/assets_object_type_order.ts (hunk 5 above)
Comment (idoamit8) [reply packages/jira-adapter/src/filters/assets/assets_object_type_order.ts:158]: I am filtering out if it is not expected :)

File: packages/jira-adapter/src/config/api_config.ts [hunk 8]
```diff
@@ -2380,7 +2380,12 @@ export const JSM_DUCKTYPE_SUPPORTED_TYPES = {
   Calendar: ['Calendar'],
   PortalSettings: ['PortalSettings'],
   SLA: ['SLA'],
-  Form: [],
+  Form: [], // being fetched by a filter.
+  AssetsSchema: [],
+  AssetsStatus: [], // being fetched by recurseInto.
+  AssetsObjectType: [], // being fetched by recurseInto.
+  AssetsObjectTypeAttribute: [], // being fetched by recurseInto.
+  [OBJECT_TYPE_ORDER_TYPE]: [], // being fetch by a filter.
```
Comment (shir-reifenberg) [inline packages/jira-adapter/src/config/api_config.ts:2388]: Is it all here to allow `exclude` / `include`? 
should we remove `OBJECT_TYPE_ORDER_TYPE` ? is there's a point to exclude it explicitly ?

File: packages/jira-adapter/src/filters/assets/assets_object_type_order.ts [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+
+/* Handles the assetsObjectTypes order inside each assets objectType
+ * by creating an InstanceElement of the assetsObjectTypes order inside the assets objectType.
+ */
+const filterCreator: FilterCreator = ({ config, client, fetchQuery }) => ({
+  name: 'assetsObjectTypeOrderFilter',
+  onFetch: async (elements: Element[]) => {
+    if (!config.fetch.enableJSM
+    || !config.fetch.enableJsmExperimental
+    || !fetchQuery.isTypeMatch(OBJECT_TYPE_ORDER_TYPE)) {
```
Comment (shir-reifenberg) [inline packages/jira-adapter/src/filters/assets/assets_object_type_order.ts:104]: maybe - 
```suggestion
    || !fetchQuery.isTypeMatch(OBJECT_TYPE_TYPE)) {
```

see my comment above 🙏

File: packages/jira-adapter/src/filters/assets/assets_object_type_order.ts [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+
+    const assetsObjectTypeInstances = elements.filter(isInstanceElement)
+      .filter(e => e.elemID.typeName === OBJECT_TYPE_TYPE)
+
+    const parentToObjectTypes = _.groupBy(
+      assetsObjectTypeInstances.filter(objectType => isReferenceExpression(objectType.value.parentObjectTypeId)),
+      objectType => objectType.value.parentObjectTypeId.elemID.getFullName()
+    )
+
+    const instanceNameToInstcne = _.keyBy(elements.filter(isInstanceElement)
```
Comment (shir-reifenberg) [inline packages/jira-adapter/src/filters/assets/assets_object_type_order.ts:116]: typo 👀 
```suggestion
    const instanceNameToInstance = _.keyBy(elements.filter(isInstanceElement)
```

File: packages/jira-adapter/src/filters/assets/assets_object_type_order.ts (hunk 5 above)
Comment (shir-reifenberg) [reply packages/jira-adapter/src/filters/assets/assets_object_type_order.ts:183]: I see you filter the after `objectTypes` (like 180) but there is no validation on the before values, no ?

**Review by shir-reifenberg — APPROVED** [review]
Looks great! added 2 suggestions but not a blocker :)

File: packages/jira-adapter/src/filters/assets/assets_object_type_order.ts (hunk 10 above)
Comment (idoamit8) [reply packages/jira-adapter/src/filters/assets/assets_object_type_order.ts:116]: OMG ! Need to reset my 34 days in a row record

**Pull request merged**